include/circt/Dialect/Pipeline/Pipeline.td

```diff
@@ -126,82 +127,181 @@ def PipelineOp : Op<Pipeline_Dialect, "pipeline", [
       return &region->front();
     }
 
+    // Returns the internal inputs of this pipeline.
+    ValueRange getInnerInputs() { return getBodyBlock()->getArguments().drop_back(1); }
+
+    // Returns the internal enable signal of this pipeline.
+    Value getInnerEnable() {
+      return getBodyBlock()->getArguments().back();
+    }
   }];
 }
 
 
-def PipelineStageOp : Op<Pipeline_Dialect, "stage", [
-    HasParent<"PipelineOp">
+def StageSeparatingOp : Op<Pipeline_Dialect, "ss", [
+    HasParent<"PipelineOp">,
+    StageLike
   ]> {
-  let summary = "Pipeline pipeline stage.";
+  let summary = "Pipeline **s**tage **s**eparating operation.";
   let description = [{
-    The `pipeline.stage` operation represents a stage separating register
-    in a pipeline. The stage does not define any explicit registers, but solely
+    The `pipeline.ss` operation represents a stage separation in a pipeline.
+    The delimiter does not define any explicit registers, but solely
     defines a cut of a dataflow graph based on its lexical position in the
     pipeline. Pipeline registers are made explicit through the register
-    materialization pass, wherein this op is replaced by
-    `pipeline.stage.register` operations.
+    materialization pass, wherein this op is replaced by `pipeline.stage`
+    operations.
   }];
 
-  let arguments = (ins I1:$when);
+  let arguments = (ins I1:$enable);
   let results = (outs I1:$valid);
 
   let assemblyFormat = [{
-    `when` $when attr-dict
+    `enable` $enable attr-dict
   }];
 
   let extraClassDeclaration = [{
     // Returns the index of this stage in the pipeline.
-    unsigned index() {
-      auto stageOps = getOperation()->getParentOfType<PipelineOp>().getOps<PipelineStageOp>();
-      return std::distance(stageOps.begin(), llvm::find(stageOps, *this));
-    }
+    unsigned index();
+
+    ValueRange getInputs() { assert(false && "Unimplemented for `pipeline.ss`"); }
```

Comment:
This feels like a bit of a code smell. If this doesn't implement the full interface, it seems like it shouldn't implement the interface, or the interface should be changed / split up. It seems like there is some representational issue here.

Reply:
Totally agree - the interface got introduced halfway into coding this up, but deserves another pass now that the actual solution is in place.